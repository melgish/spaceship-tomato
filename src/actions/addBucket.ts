'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddBucket, addBucketSchema } from '@/schemas/addBucket.schema'

/**
 * Add server action to a new bucket to the database.
 * @see ../schemas/addBucket.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addBucket(payload: AddBucket) {
  try {
    const data = await addBucketSchema.parseAsync(payload)
    const created = await prisma.bucket.create({
      data: { ...data, plantId: data.plantId || null },
    })
    revalidatePath('/buckets')
    return { ok: true, message: 'Bucket created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create bucket' } as const
  }
}
