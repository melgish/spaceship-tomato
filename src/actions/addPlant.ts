'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddPlant, addPlantSchema } from '@/schemas/addPlant.schema'

/**
 * Add server action to a new plant to the database.
 * @see ../schemas/addPlant.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addPlant(payload: AddPlant) {
  try {
    const data = await addPlantSchema.parseAsync(payload)
    const created = await prisma.plant.create({ data })
    revalidatePath('/plants')
    return { ok: true, message: 'Plant created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create plant' } as const
  }
}
