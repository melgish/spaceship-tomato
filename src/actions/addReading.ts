'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddReading, addReadingSchema } from '@/schemas/addReading.schema'

/**
 * Add server action to a new plant to the database.
 * @see ../schemas/addReading.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addReading(payload: AddReading) {
  try {
    const data = await addReadingSchema.parseAsync(payload)
    const created = await prisma.reading.create({ data })
    revalidatePath('/readings')
    return { ok: true, message: 'Reading created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create reading' } as const
  }
}
