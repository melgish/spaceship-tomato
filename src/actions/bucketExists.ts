'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that a bucket exists in the database.
 * @param id Id of bucket to search for.
 * @returns true if bucket exists, otherwise false.
 */
export async function bucketExists(id: number = 0) {
  return id > 0 && (await prisma.bucket.count({ where: { id } })) > 0
}
