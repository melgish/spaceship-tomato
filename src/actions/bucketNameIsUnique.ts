'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that the name of a bucket is unique.
 * @param name Name of the bucket to test.
 * @param id Optional id of bucket to exclude when editing an existing bucket.
 * @returns true if the name exists on another bucket, otherwise false.
 */
export async function bucketNameIsUnique(name: string, id: number = 0) {
  const where = { name, id: (id > 0 && { not: id }) || undefined }
  return (await prisma.bucket.count({ where })) === 0
}
