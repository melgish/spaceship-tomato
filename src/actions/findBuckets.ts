'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to fetch all of buckets from the database.
 * @returns An array of buckets. The returned data will include plant info if
 * the bucket contains a plant.
 */
export async function findBuckets() {
  return await prisma.bucket.findMany({
    include: { plant: true },
    orderBy: [{ name: 'asc' }],
  })
}

export type Buckets = Awaited<ReturnType<typeof findBuckets>>
export type Bucket = Buckets[number]
