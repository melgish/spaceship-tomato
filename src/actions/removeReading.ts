'use server'
import { prisma } from '@/lib/prisma'
import { revalidatePath } from 'next/cache'

/**
 * A server action to remove a single reading from the system.
 * @param id Id of the reading to remove.
 * @returns an Ok result if the identified reading was removed, otherwise an
 *  Err result.
 */
export async function removeReading(id: number) {
  try {
    await prisma.reading.delete({ where: { id }, select: { id: true } })
    revalidatePath('/readings')
    return { ok: true, message: 'Reading deleted', data: id } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to remove reading' } as const
  }
}
