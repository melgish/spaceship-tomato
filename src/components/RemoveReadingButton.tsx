'use client'
import { removeReading } from '@/actions/removeReading'
import { notify } from '@/lib/notify'
import RemoveButton from './RemoveButton'

type Props = Readonly<{ row?: { id: number } }>

export default function RemoveReadingButton({ row }: Props) {
  if (row?.id) {
    return (
      <RemoveButton
        onClick={() => {
          // Resolve was required to make test work even when mocked?
          Promise.resolve(removeReading(row.id)).then(notify)
        }}
      />
    )
  }
}
