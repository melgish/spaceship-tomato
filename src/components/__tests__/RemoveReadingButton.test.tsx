import { describe, expect, it, vi } from 'vitest'
import { act, fireEvent, render, screen } from '@testing-library/react'
import RemoveReadingButton from '@/components/RemoveReadingButton'
import { removeReading } from '@/actions/removeReading'

const data = { id: 1 }

vi.mock('@/actions/removeReading', () => ({
  removeReading: vi.fn().mockResolvedValue({
    ok: true,
    message: 'Bucket deleted',
    data: 1,
  }),
}))
vi.mock('@/lib/notify', () => ({ notify: vi.fn() }))

describe('when row is not supplied', () => {
  it('should not render', () => {
    const { container } = render(<RemoveReadingButton />)

    expect(container).toMatchSnapshot()
  })
})

describe('when row is supplied', () => {
  it('should render', () => {
    const { container } = render(<RemoveReadingButton row={data} />)

    expect(container).toMatchSnapshot()
  })
})

describe('when button is clicked', () => {
  it('should invoke removeReading', async () => {
    render(<RemoveReadingButton row={data} />)

    await act(() => fireEvent.click(screen.getByRole('button')))

    expect(removeReading).toBeCalled()
  })
})
