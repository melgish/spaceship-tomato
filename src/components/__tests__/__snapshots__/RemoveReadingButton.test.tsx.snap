// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`when row is not supplied > should not render 1`] = `<div />`;

exports[`when row is supplied > should render 1`] = `
<div>
  <button
    class="px-1 py-0 text-red-800 font-bold border-red-800 text-sm"
    type="button"
  >
    Remove
  </button>
</div>
`;
